backend/tournesol/views/inconsistencies.py

```diff
@@ -8,12 +8,188 @@
+
+    def _fill_graph_parameters(self, comparisons: list):
+        """
+        Fill the set containing all the entities.
+
+        The incoming and outgoing connections sets are used to search the cycles.
+        all_connections is only used to count all the comparison trios.
+
+        outgoing_connections: Dictionary of sets for each entity.
+                              Each element in the sets is a preferred entity.
+        incoming_connections: Idem but each element in the sets is a worse rated entity.
+
+        all_connections: all the edges (also including comparisons with a score of 0 as edges)
+        """
+        self.all_entities = set()
+        self.outgoing_connections = {}
+        self.incoming_connections = {}
+
+        # all the edges (also including comparisons with a score of 0 as edges)
+        self.all_connections = {}
+
+        criterion_comparisons = list(
+            filter(lambda d: d["criteria"] == self.criterion, comparisons)
+        )
+
+        for comparison in criterion_comparisons:
+            entity_1 = comparison["comparison__entity_1__uid"]
+            entity_2 = comparison["comparison__entity_2__uid"]
+
+            if entity_1 not in self.all_entities:
+                self.all_entities.add(entity_1)
+                self.incoming_connections[entity_1] = set()
+                self.outgoing_connections[entity_1] = set()
+                self.all_connections[entity_1] = set()
+
+            if entity_2 not in self.all_entities:
+                self.all_entities.add(entity_2)
+                self.incoming_connections[entity_2] = set()
+                self.outgoing_connections[entity_2] = set()
+                self.all_connections[entity_2] = set()
+
+            if comparison["score"] > 0:
+                # Entity_2 is preferred
+                self.incoming_connections[entity_2].add(entity_1)
+                self.outgoing_connections[entity_1].add(entity_2)
+
+            elif comparison["score"] < 0:
+                # Entity_1 is preferred
+                self.incoming_connections[entity_1].add(entity_2)
+                self.outgoing_connections[entity_2].add(entity_1)
+            else:
+                continue
+
+            self.all_connections[entity_1].add(entity_2)
+            self.all_connections[entity_2].add(entity_1)
+
+    def _count_cycles_and_comparison_trios(self):
+        """
+        Count the cycles of length 3 and the comparison trios.
+        Add the cycles found to the cycles list.
+
+        The cycles returned are in the order 1, 2, 3 with 1 > 2 > 3 > 1.
+        """
+        cycles_count = 0
+        comparison_trios_count = 0
+        for entity_1 in self.all_entities:
+            # Search the cycles
+            for entity_2 in self.outgoing_connections[entity_1].copy():
+                for entity_3 in self.incoming_connections[entity_1] & \
+                                self.outgoing_connections[entity_2]:
```

Comment:
The code would feel more readable to me with something like this:
```
                for entity_3 in self.outgoing_connections[entity_2]:
                    if entity_1 in self.outgoing_connections[entity_3]:
```
I don't think it reduces performance, does it?
Maybe we don't need to keep track of incoming edges with that change.

Reply:
Actually, the & operator between sets is very efficient, especially when the sets have few elements in common. I made some experiments on python.

Reply:
Honestly, the sets in incoming_connections and outgoing_connections must be pretty small in practice, so I'm not sure that the gain is significant, but I prefer to keep it like this.

Reply:
Agreed! Thanks